cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java

```diff
@@ -85,52 +84,41 @@
+  public ETLMapReduce(BatchPhaseSpec phaseSpec) {
+    this.phaseSpec = phaseSpec;
   }
 
   @Override
   public void configure() {
-    setName(NAME);
-    setDescription("MapReduce Driver for ETL Batch Applications");
-
-    PipelineRegisterer pipelineRegisterer = new PipelineRegisterer(getConfigurer(), "batch");
-
-    PipelinePhase pipeline =
-      pipelineRegisterer.registerPlugins(
-        config, TimePartitionedFileSet.class,
-        FileSetProperties.builder()
-          .setInputFormat(AvroKeyInputFormat.class)
-          .setOutputFormat(AvroKeyOutputFormat.class)
-          .setEnableExploreOnCreate(true)
-          .setSerDe("org.apache.hadoop.hive.serde2.avro.AvroSerDe")
-          .setExploreInputFormat("org.apache.hadoop.hive.ql.io.avro.AvroContainerInputFormat")
-          .setExploreOutputFormat("org.apache.hadoop.hive.ql.io.avro.AvroContainerOutputFormat")
-          .setTableProperty("avro.schema.literal", Constants.ERROR_SCHEMA.toString())
-          .build(), true);
-
-    Resources resources = config.getResources();
-    if (resources != null) {
-      setMapperResources(resources);
+    setName(phaseSpec.getPhaseName());
+
+    StringBuilder description = new StringBuilder("DataFlow MapReduce phase executor. Sources");
+
+    for (String sourceName : phaseSpec.getPhase().getSources()) {
+      description.append(" '").append(sourceName).append("'");
     }
-    Resources driverResources = config.getDriverResources();
-    if (driverResources != null) {
-      setDriverResources(driverResources);
+    description.append(" to sinks");
+
+    for (StageInfo sinkInfo : phaseSpec.getPhase().getStagesOfType(BatchSink.PLUGIN_TYPE)) {
+      description.append(" '").append(sinkInfo.getName()).append("'");
     }
+    setDescription(description.toString());
 
-    // following should never happen
-    Preconditions.checkNotNull(pipeline, "Pipeline is null");
-    Preconditions.checkNotNull(pipeline.getSinks(), "Sinks could not be found in program properties");
-    // empty transform list is created during pipeline register
-    Preconditions.checkNotNull(pipeline.getTransforms());
-    Preconditions.checkNotNull(pipeline.getConnections(), "Connections could not be found in program properties");
+    setMapperResources(phaseSpec.getResources());
+    setDriverResources(phaseSpec.getResources());
+
+    if (phaseSpec.getPhase().getSources().size() != 1) {
+      throw new IllegalArgumentException("Pipeline must contain exactly one source.");
+    }
+    if (phaseSpec.getPhase().getSinks().isEmpty()) {
+      throw new IllegalArgumentException("Pipeline must contain at least one sink.");
+    }
 
     // add source, sink, transform ids to the properties. These are needed at runtime to instantiate the plugins
     Map<String, String> properties = new HashMap<>();
-    properties.put(Constants.PIPELINEID, GSON.toJson(pipeline));
-    properties.put(Constants.STAGE_LOGGING_ENABLED, String.valueOf(config.isStageLoggingEnabled()));
```

Comment:
I put everything (including stage logging) in the BatchPhaseSpec since I didn't want to serialize and deserialize 5 different properties